services/ingestion/block_tracking_subscriber.go

```diff
@@ -0,0 +1,303 @@
+		if r.client.IsPastSpork(r.height) {
+			r.logger.Info().
+				Uint64("height", r.height).
+				Msg("height found in previous spork, starting to backfill")
+
+			// backfill all the missed events, handling of context cancellation is done by the producer
+			for ev := range r.backfill(ctx, r.height) {
+				eventsChan <- ev
+
+				if ev.Err != nil {
+					return
+				}
+
+				// keep updating height, so after we are done back-filling
+				// it will be at the first height in the current spork
+				r.height = ev.Events.CadenceHeight()
+			}
+
+			// after back-filling is done, increment height by one,
+			// so we start with the height in the current spork
+			r.height = r.height + 1
+		}
+
+		r.logger.Info().
+			Uint64("next-height", r.height).
+			Msg("backfilling done, subscribe for live data")
+
+		// subscribe in the current spork, handling of context cancellation is done by the producer
+		for ev := range r.subscribe(ctx, r.height) {
+			eventsChan <- ev
+		}
+
+		r.logger.Warn().Msg("ended subscription for events")
+	}()
+
+	return eventsChan
+}
+
+// subscribe to events by the provided height and handle any errors.
+//
+// Subscribing to EVM specific events and handle any disconnection errors
+// as well as context cancellations.
+func (r *RPCBlockTrackingSubscriber) subscribe(ctx context.Context, height uint64) <-chan models.BlockEvents {
+	eventsChan := make(chan models.BlockEvents)
+
+	var blockHeadersChan <-chan flow.BlockHeader
+	var errChan <-chan error
+
+	lastReceivedHeight := height
+	connect := func(height uint64) error {
+		var err error
+		blockHeadersChan, errChan, err = r.client.SubscribeBlockHeadersFromStartHeight(
+			ctx,
+			height,
+			flow.BlockStatusFinalized,
+		)
+		return err
+	}
+
+	if err := connect(lastReceivedHeight); err != nil {
+		eventsChan <- models.NewBlockEventsError(
+			fmt.Errorf(
+				"failed to subscribe for finalized block headers on height: %d, with: %w",
+				height,
+				err,
+			),
+		)
+		close(eventsChan)
+		return eventsChan
+	}
+
+	go func() {
+		defer func() {
+			close(eventsChan)
+		}()
+
+		for ctx.Err() == nil {
+			select {
+			case <-ctx.Done():
+				r.logger.Info().Msg("event ingestion received done signal")
+				return
+
+			case blockHeader, ok := <-blockHeadersChan:
+				if !ok {
+					// typically we receive an error in the errChan before the channels are closes
+					var err error
+					err = errs.ErrDisconnected
+					if ctx.Err() != nil {
+						err = ctx.Err()
+					}
+					eventsChan <- models.NewBlockEventsError(err)
+					return
+				}
+
+				blockEvents, err := r.evmEventsForBlock(ctx, blockHeader)
+				if err != nil {
+					eventsChan <- models.NewBlockEventsError(err)
+					return
+				}
+
+				evmEvents := models.NewSingleBlockEvents(blockEvents)
+				// if events contain an error, or we are in a recovery mode
+				if evmEvents.Err != nil || r.recovery {
+					evmEvents = r.recover(ctx, blockEvents, evmEvents.Err)
```

Comment:
I think currently you might not get to here since the error will already happen on `evmEventsForBlock`